sw/device/lib/drivers/rv_plic/rv_plic.c

```diff
@@ -34,337 +81,417 @@ typedef struct plic_reg_info {
 }
 
-/**
- * Get an IRQ source specific Level/Edge register info.
- */
-static plic_reg_info_t plic_irq_trigger_type_reg_info(dif_plic_irq_id_t irq) {
-  ptrdiff_t offset = plic_offset_from_reg0(irq);
-  return (plic_reg_info_t){
-      .offset = RV_PLIC_LE_0_REG_OFFSET + offset,
-      .bit_index = plic_irq_bit_index(irq),
-  };
-}
 
-/**
- * Get an IRQ source specific Interrupt Pending register info.
- */
-static plic_reg_info_t plic_irq_pending_reg_info(dif_plic_irq_id_t irq) {
-  ptrdiff_t offset = plic_offset_from_reg0(irq);
-  return (plic_reg_info_t){
-      .offset = RV_PLIC_IP_0_REG_OFFSET + offset,
-      .bit_index = plic_irq_bit_index(irq),
-  };
-}
+void handler_irq_external(void)
+{
+  plic_irq_id_t int_id;
+  plic_result_t res = plic_irq_claim(0, &int_id);
+  irq_sources_t type = plic_get_irq_src_type(int_id);
+
+  switch (type)
+  {
+  case IRQ_UART_SRC: 
+    handler_irq_uart();
+    external_intr_flag = 1;
+    break;
+  
+  case IRQ_GPIO_SRC:
+    handler_irq_gpio();
+    external_intr_flag = 1;
+    break;
+
+  case IRQ_I2C_SRC:
+    handler_irq_i2c();
+    external_intr_flag = 1;
+    break;
+
+  case IRQ_SPI_SRC:
+    handler_irq_spi();
+    external_intr_flag = 1;
+    break;
+
+  case IRQ_BAD:
+    printf("Bad INT source!\n");
+    break;
+
+  default:
+    break;
+  }
 
-/**
- * Get a PRIO register offset (PRIO0, PRIO1, ...) from an IRQ source ID.
- *
- * There is one PRIO register per IRQ source, this function calculates the IRQ
- * source specific PRIO register offset.
- */
-static ptrdiff_t plic_priority_reg_offset(dif_plic_irq_id_t irq) {
-  ptrdiff_t offset = irq * sizeof(uint32_t);
-  return RV_PLIC_PRIO0_REG_OFFSET + offset;
+  plic_irq_complete(&int_id);
 }
 
-/**
- * Reset the requested PLIC peripheral.
- *
- * This function resets all the relevant PLIC registers, apart from the CC
- * register. There is no reliable way of knowing the ID of an IRQ that has
- * claimed the CC register, so we assume that the previous "owner" of the
- * resource has cleared/completed the CC access.
- */
-static void plic_reset(const dif_plic_t *plic) {
-  // Clear all of the Level/Edge registers.
-  for (int i = 0; i < RV_PLIC_LE_MULTIREG_COUNT; ++i) {
-    ptrdiff_t offset = RV_PLIC_LE_0_REG_OFFSET + (i * sizeof(uint32_t));
-    mmio_region_write32(plic->params.base_addr, offset, 0);
-  }
 
-  // Clear all of the priority registers.
-  for (int i = 0; i < RV_PLIC_PARAM_NUM_SRC; ++i) {
-    ptrdiff_t offset = plic_priority_reg_offset(i);
-    mmio_region_write32(plic->params.base_addr, offset, 0);
+/*!
+  Resets relevant registers of the PLIC (Level/Edge, 
+  priority, target, threshold, interrupts).
+  It writes the default value 0 in them and read back
+  the value into the proper register struct
+*/
+plic_result_t plic_Init(void)
+{
+  
+  /* Clears all the Level/Edge registers */ 
+  for(size_t i=0; i<RV_PLIC_LE_MULTIREG_COUNT; i++)
+  {
+    uint32_t *dest = (&rv_plic_peri->IP0 + i);
+    *dest = 0;
+
+    if (*dest != 0) 
```

Comment:
MINOR: we need to discuss this, basically, you are assuming that there could be a problem where the register cannot be set, right? I mean... can this really happen?